controller/proxy-injector/webhook.go

```diff
@@ -212,14 +212,19 @@ func (w *Webhook) inject(request *admissionv1beta1.AdmissionRequest) (*admission
 	return admissionResponse, nil
 }
 
-// ignore determines whether or not the given deployment should be injected.
-// A deployment is ignored if:
+// shouldInject determines whether or not the given deployment should be
+// injected. A deployment should be injected if it does not already contain
+// any known sidecars, and:
 // - the deployment's namespace has the linkerd.io/inject annotation set to
-//   "disabled", and the deployment's pod spec does not have the
-//   linkerd.io/inject annotation set to "enabled"; or
+//   "enabled", and the deployment's pod spec does not have the
+//   linkerd.io/inject annotation set to "disabled"; or
 // - the deployment's pod spec has the linkerd.io/inject annotation set to
-//   "disabled"
-func (w *Webhook) ignore(ns string, deployment *appsv1.Deployment) (bool, error) {
+//   "enabled"
+func (w *Webhook) shouldInject(ns string, deployment *appsv1.Deployment) (bool, error) {
+	if healthcheck.HasExistingSidecars(&deployment.Spec.Template.Spec) {
+		return false, nil
```

Comment:
this may be out of scope for this PR, but i think putting the `HasExistingSidecars` first changes the semantics a bit? this would prevent re-injection of pod following a Linkerd control-plane upgrade.

if we do want to support the upgrade case, perhaps we need to modify `HasExistingSidecars` to distinguish between a `linkerd-proxy` and other proxies?

Reply:
We briefly did that in the `uninject before inject` PR but then rolled that back, the reason being already injected pods might contained special settings that might get lost during a proxy upgrade
https://github.com/linkerd/linkerd2/pull/2089#discussion_r248463876

Reply:
Good points. At the moment, this webhook is only triggered when new deployments are created. Yes, as part of the story for data plane upgrade in an auto-inject env, `HasExistingSidecars` will need to be a bit smarter.